packages/datadog-instrumentations/src/vitest.js

```diff
@@ -0,0 +1,282 @@
+    return 'skip'
+  }
+  return 'pass'
+}
+
+// eslint-disable-next-line
+// From https://github.com/vitest-dev/vitest/blob/51c04e2f44d91322b334f8ccbcdb368facc3f8ec/packages/runner/src/run.ts#L243-L250
+function getVitestTestStatus (test, retryCount) {
+  if (test.result.state !== 'fail') {
+    if (!test.repeats) {
+      return 'pass'
+    } else if (test.repeats && (test.retry ?? 0) === retryCount) {
+      return 'pass'
+    }
+  }
+  return 'fail'
+}
+
+function getTypeTasks (fileTasks, type = 'test') {
+  const typeTasks = []
+
+  function getTasks (tasks) {
+    for (const task of tasks) {
+      if (task.type === type) {
+        typeTasks.push(task)
+      } else if (task.tasks) {
+        getTasks(task.tasks)
+      }
+    }
+  }
+
+  getTasks(fileTasks)
+
+  return typeTasks
+}
+
+function getTestName (task) {
+  let testName = task.name
+  let currentTask = task.suite
+
+  while (currentTask) {
+    if (currentTask.name) {
+      testName = `${currentTask.name} ${testName}`
+    }
+    currentTask = currentTask.suite
+  }
+
+  return testName
+}
+
+addHook({
+  name: 'vitest',
+  versions: ['>=1.6.0'],
+  file: 'dist/runners.js'
+}, (vitestPackage) => {
+  const { VitestTestRunner } = vitestPackage
+  // test start (only tests that are not marked as skip or todo)
+  shimmer.wrap(VitestTestRunner.prototype, 'onBeforeTryTask', onBeforeTryTask => async function (task) {
+    if (!testStartCh.hasSubscribers) {
+      return onBeforeTryTask.apply(this, arguments)
+    }
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    taskToAsync.set(task, asyncResource)
+
+    asyncResource.runInAsyncScope(() => {
+      testStartCh.publish({ testName: getTestName(task), testSuiteAbsolutePath: task.suite.file.filepath })
+    })
+    return onBeforeTryTask.apply(this, arguments)
+  })
+
+  // test finish (only passed tests)
+  shimmer.wrap(VitestTestRunner.prototype, 'onAfterTryTask', onAfterTryTask =>
+    async function (task, { retry: retryCount }) {
+      if (!testFinishTimeCh.hasSubscribers) {
+        return onAfterTryTask.apply(this, arguments)
+      }
+      const result = await onAfterTryTask.apply(this, arguments)
+
+      const status = getVitestTestStatus(task, retryCount)
+      const asyncResource = taskToAsync.get(task)
+
+      if (asyncResource) {
+        // We don't finish here because the test might fail in a later hook
+        asyncResource.runInAsyncScope(() => {
+          testFinishTimeCh.publish({ status, task })
+        })
+      }
+
+      return result
+    })
+
+  return vitestPackage
+})
+
+addHook({
+  name: 'vitest',
+  versions: ['>=1.6.0'],
+  filePattern: 'dist/vendor/index.*'
+}, (vitestPackage) => {
+  // there are multiple index* files so we have to check the exported values
+  if (!isReporterPackage(vitestPackage)) {
+    return vitestPackage
+  }
+  shimmer.wrap(vitestPackage.B.prototype, 'sort', sort => async function () {
+    if (!testSessionFinishCh.hasSubscribers) {
```

Comment:
vitest plugin couldn't be enabled after the sort method is executed?

Reply:
I guess, but by this time the session has already started, so if `testSessionFinishCh` has no subscribers by now, it's not going to matter.

Did you ask this because we could move this check to the `this.ctx.exit` wrap?